presto-main/src/main/java/io/prestosql/spiller/AesSpillCipher.java

```diff
@@ -0,0 +1,172 @@
+        this.ivBytes = encryptSizer.getIV().length;
+    }
+
+    @Override
+    public int encryptedMaxLength(int inputLength)
+    {
+        checkArgument(inputLength >= 0, "inputLength must be >= 0, inputLength=%s", inputLength);
+        return ivBytes + throwCipherClosedIfNull(encryptSizer).getOutputSize(inputLength);
+    }
+
+    @Override
+    public int decryptedMaxLength(int encryptedLength)
+    {
+        checkArgument(encryptedLength >= ivBytes, "encryptedLength must be >= %s", ivBytes);
+        return encryptedLength - ivBytes;
+    }
+
+    @Override
+    public int encrypt(byte[] data, int inputOffset, int length, byte[] destination, int destinationOffset)
+    {
+        checkArgument(data.length - inputOffset >= length, "data buffer too small for length argument");
+        checkArgument(destination.length - destinationOffset >= encryptedMaxLength(length), "destination buffer too small for encrypted output");
+        Cipher cipher = createEncryptCipher();
+        System.arraycopy(cipher.getIV(), 0, destination, destinationOffset, ivBytes);
+        try {
+            return ivBytes + cipher.doFinal(data, inputOffset, length, destination, destinationOffset + ivBytes);
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failed to encrypt data: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public int decrypt(byte[] encryptedData, int inputOffset, int length, byte[] destination, int destinationOffset)
+    {
+        checkArgument(encryptedData.length - inputOffset >= length, "encryptedData too small for length argument");
+        checkArgument(destination.length - destinationOffset >= decryptedMaxLength(length), "destination buffer too small for decrypted output");
+        Cipher cipher = createDecryptCipher(new IvParameterSpec(encryptedData, inputOffset, ivBytes));
+        try {
+            return cipher.doFinal(encryptedData, inputOffset + ivBytes, length - ivBytes, destination, destinationOffset);
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Cannot decrypt previously encrypted data: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        /*
+         * Setting the {@link AesSpillCipher#key} to null allows the key to be reclaimed by GC at the time of
+         * destruction, even if the {@link AesSpillCipher} itself is still reachable for longer. When the key
+         * is null, subsequent encrypt / decrypt operations will fail and can prevent accidental use beyond the
+         * intended lifespan of the {@link AesSpillCipher}.
+         */
+        this.key = null;
+        this.encryptSizer = null;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this).add("closed", key == null).toString();
+    }
+
+    private Cipher createEncryptCipher()
```

Comment:
I feel like there is one too many layers here.  What do you think about having only `static Cipher createEncryptCipher(SecretKey key)` and `static Cipher createDecryptCipher(SecretKey key, IvParameterSpec iv)`?

Reply:
Sure, that works too. Will do.